module/VuFind/src/VuFindTest/Feature/WithConsecutiveTrait.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+/**
+ * Trait for setting consecutive test expectations.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Villanova University 2024.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:testing:unit_tests Wiki
+ */
+
+namespace VuFindTest\Feature;
+
+use PHPUnit\Framework\Constraint\Constraint;
+use PHPUnit\Framework\MockObject\Builder\InvocationStubber;
+use PHPUnit\Framework\MockObject\MockObject;
+
+use function count;
+use function func_get_args;
+use function is_array;
+
+/**
+ * Trait for setting consecutive test expectations.
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:testing:unit_tests Wiki
+ */
+trait WithConsecutiveTrait
+{
+    /**
+     * Expect consecutive calls to a mock.
+     *
+     * @param MockObject $mock          Mock object
+     * @param string     $method        Method expecting calls
+     * @param array      $expectedCalls Expected input parameters
+     * @param mixed      $returnValues  Return values to mock (either an array indexed parallel to
+     * $expectedCalls to return different values, or a single value to always return the same thing)
+     *
+     * @return InvocationStubber
+     */
+    protected function expectConsecutiveCalls(
+        MockObject $mock,
+        string $method,
+        array $expectedCalls,
+        mixed $returnValues = null
+    ): InvocationStubber {
+        $matcher = $this->exactly(count($expectedCalls));
+        $callback = function () use ($matcher, $expectedCalls, $returnValues) {
+            $index = $matcher->getInvocationCount() - 1;
+            $expectedArgs = $expectedCalls[$index];
```

Comment:
Should check if the given index exists in $expectedCalls, right?

Reply:
My thinking was that our exactly assertion based on the count of expected calls should guarantee this, but maybe it's still worth being more strict. What do you think?

Reply:
Depends on whether it throws immediately when the expected count is exceeded. Do you have insight? Maybe simply null coalescing would be the way?

Reply:
I'll try it later this morning, see how it behaves, and adjust accordingly. Thanks!

Reply:
It does appear that the exactly assertion triggers a failure before the code would attempt to access an undefined value. However, since PHPUnit internals seem to change frequently without warning, I added a null coalesce to an empty array just in case it's needed in future.